### PR TITLE
Implementing the extragalactic model from Websky

### PR DESCRIPTION
**Pull request by msyriac**
This is WIP and currently only does CIB. We will add tSZ, kSZ, kappa and a unifying class/function with PrecomputedAlms. Docs and unit tests also need to be added before merging.

* [x] use the internal interpolating component and implement download maps on the fly (@zonca)
* [x] implement unit tests (@msyriac)
* [x] add docs (@marcelo-alvarez)

### CODE REVIEW

**Comment by zonca** [comment]
it looks good, can you add a unit test?

@marcelo-alvarez can you add some documentation in https://github.com/simonsobs/so_pysm_models/blob/master/docs/models.rst, a summary of WebSky with references and description of the CIB component?

we can finalize this pull request then add other components later.

**Comment by zonca** [comment]
also, what is the path at NERSC with those maps?

**Comment by marcelo-alvarez** [comment]
> also, what is the path at NERSC with those maps?

/project/projectdirs/sobs/v4_sims/mbs/websky/data/v0.2

there are 18 cib fits files at 4096 there, of intensity in MJy / sr at

[27, 39, 93, 145, 225, 280] +/- 1 GHz

i.e. 'cib_ns4096_nu'+str(int(frequency_in_GHz)).zfill(4)+'.fits'

**Comment by zonca** [comment]
thanks @marcelo-alvarez ,

* can we post those map publicly? so that people can run simulations outside of NERSC (we are doing this for the foreground models)
* in the documentation please also explain the reason of the +/- frequency spacing

**Comment by marcelo-alvarez** [comment]
> thanks @marcelo-alvarez ,
> 
> * can we post those map publicly? so that people can run simulations outside of NERSC (we are doing this for the foreground models)
yes, by all means please do.

> * in the documentation please also explain the reason of the +/- frequency spacing
I will. this is because we are first generating sims with delta functions at a specific set of band centers so it makes sense to have the inputs closely spaced near those bands, initially. future versions will have more cib frequencies, but only after we have implemented a full set of map based simulations at those bands that include correlated lensing and SZ components as well.

**Comment by zonca** [comment]
ok, maps are available at http://portal.nersc.gov/project/cmb/so_pysm_models_data/websky/0.2/cib/

I created a checklist at the top of the pull request with things to do

**Comment by zonca** [comment]
I rebased the branch on the current `master` so I have the new `InterpolatingComponent` available, @marcelo-alvarez @msyriac please delete the old branch and use the new one.

**Comment by zonca** [comment]
ok, I implemented downloading of the maps remotely and unit tests. @msyriac would you review the code within a week?

@marcelo-alvarez once we have the docs we can merge this.

**Comment by msyriac** [comment]
I'll aim to finish reviewing this week.

**Comment by marcelo-alvarez** [comment]
> I'll aim to finish reviewing this week.

As I understand we want to finalize this PR before moving on to kSZ and tSZ. I will try to read in with the master branch of so_pysm_models. @zonca I don't see the documentation I added on the master branch is that intentional?

**Comment by zonca** [comment]
Yes, we need to merge this, but we need to wait for mat review

**Comment by msyriac** [comment]
Working on some edits to make it compatible with websky v0.3

**Comment by marcelo-alvarez** [comment]
The CIB looks good to me. Shall I go ahead and add kSZ and tSZ here or make a new PR for that?

**Comment by msyriac** [comment]
I recommend adding it here since this PR is supposed to be for "the extragalactic model from websky". tSZ, kSZ and kappa (the last for those who want to do validation).

**Comment by marcelo-alvarez** [comment]
Ok, I will take a shot at kSZ and tSZ now.

File: docs/gettingstarted.rst [hunk 1]
```diff
@@ -54,3 +54,4 @@ see example notebooks:
 * `Example Gaussian Synchrotron <https://gist.github.com/zonca/51a6fa9763106c78813f964a4b88f0fc>`_
 * `Example Gaussian Dust <https://gist.github.com/zonca/4ddb5e384cb34f8a2945c041d13e9428>`_
 * `Example InterpolatingComponent <https://gist.github.com/zonca/08751497b040ec9d62ff5175573c786e>`_
+* `Example Websky Extragalactic <https://gist.github.com/marcelo-alvarez/b13afc4a761c61334d11b5eeae953923>`_
```
Comment (zonca) [inline docs/gettingstarted.rst]: thanks for providing an example notebook!

File: docs/models.rst [hunk 2]
```diff
@@ -184,10 +184,35 @@ For example:
...
 
-`Full example notebook <https://gist.github.com/zonca/08751497b040ec9d62ff5175573c786e>`
+`Full example notebook <https://gist.github.com/zonca/08751497b040ec9d62ff5175573c786e>`_
+
+WebSky
+======
+
+The Websky suite of simulated extragalactic component maps, determined from large scale structure light cone realizations and  based on Lagrangian perturbation theory, Peak Patch Lagrangian halo finding, and modeling of SZ and CIB effects, can be read into PySM as precomputed external fits files using InterpolatingComponent.  More information on the Peak Patch halo finding method can be found in `Stein, Alvarez, and Bond (2018) <https://arxiv.org/abs/1810.07727>`_, and selected maps and halo catalogs are available from the `Websky website <http://mocks.cita.utoronto.ca/websky>`_. For a description of the models as implemented for Simons Observatory, see the `SO Websky model repository <https://github.com/simonsobs/websky_model>`_.
+
+Websky maps generated as input for PySM are described below. Currently, cosmic infrared background at SO frequencies are available, with tSZ and kSZ coming soon.
```
Comment (zonca) [inline docs/models.rst]: `tSZ` and `kSZ` are available, can @marcelo-alvarez update the docs?

File: docs/models.rst [hunk 3]
```diff
@@ -184,10 +184,35 @@ For example:
...
+
+The Planck (2013) CIB halo model is used, along with a halo occupation distribution. More details can be found `here <https://github.com/simonsobs/websky_model>`_. 
+
+The current version of the maps are of intensity in units of MJy/Sr with filename convention ``cib_ns[NSIDE]_nu[FREQ].fits`` e.g. ``cib_ns4096_nu0027.fits`` is the map of CIB intensity at 27 GHz and will be used by :py:meth:`.InterpolatingComponent` at that frequency, and can be found on NERSC at `/project/projectdirs/sobs/v4_sims/mbs/websky/data/v0.2`. There are 18 fits files at :math:`N_{side}`=4096 at frequencies [27, 39, 93, 145, 225, 280] +/- 1 GHz. These intensities were selected because in order to be able to interpolate accurately at the 6 frequencies of interest with as few maps as possible. More frequencies will be made available after a full set of map based simulations at those bands that include correlated lensing and SZ components has been implemented.
+
+**Thermal SZ Effect**
+
+Provided is a map of the Compton-y parameter and is based on Battaglia et al. (2012) pressure profiles, more details `here <https://github.com/simonsobs/websky_model>`_. Not yet implemented in PySM.
```
Comment (zonca) [inline docs/models.rst]: @marcelo-alvarez update here as well

File: docs/models.rst [hunk 4]
```diff
@@ -184,10 +184,35 @@ For example:
...
+
+**Kinetic SZ Effect**
+
+Provided is a map of the temperature fluctuation due to line of sight peculiar velocities of electrons along the line of sight. Electrons are assumed to follow an NFW profile interior to halos and second order LPT outside, more details `here <https://github.com/simonsobs/websky_model>`_. Not yet implemented in PySM.
```
Comment (zonca) [inline docs/models.rst]: @marcelo-alvarez can you expand NFW and LPT?

File: so_pysm_models/extragalactic.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
+from . import InterpolatingComponent
+from . import utils
+import pysm
+import numpy as np
+import healpy as hp
+
+
+def y2uK_CMB(nu):
```
Comment (zonca) [inline so_pysm_models/extragalactic.py]: @marcelo-alvarez can you add docstring with detail about this conversion?

File: so_pysm_models/tests/test_extragalactic.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    hp.write_map(path / "ksz.fits", np.ones(shape, dtype=np.float32))
+    hp.write_map(path / "cib_0100.fits", np.ones(shape, dtype=np.float32))
+
+    ksz = WebSkySZ("0.3", sz_type="kinetic", target_nside=nside)
+
+    ksz_map = ksz.signal(nu=100)
+    np.testing.assert_allclose(
+        np.ones(ksz_map.shape)*0.777228, ksz_map, rtol=1e-4
+    )
+
```
Comment (zonca) [inline so_pysm_models/tests/test_extragalactic.py]: @marcelo-alvarez I have added 2 unit tests for `ksz` and `tsz`.
I do not want to actually download the real maps, so I create nside=4 maps full of ones and I trick  the classes into loading those.
Then I executed the code once, got the result, for example 0.777228 in the `ksz` case and make sure that future changes do not mess this up.
Is this ok? do you have in mind a better way to test this?

File: docs/models.rst [hunk 7]
```diff
@@ -184,10 +184,35 @@ For example:
...
+
+Websky maps generated as input for PySM are described below. Currently, cosmic infrared background at SO frequencies are available, with tSZ and kSZ coming soon.
+
+**Cosmic Infrared Background**
+
+The Planck (2013) CIB halo model is used, along with a halo occupation distribution. More details can be found `here <https://github.com/simonsobs/websky_model>`_. 
+
+The current version of the maps are of intensity in units of MJy/Sr with filename convention ``cib_ns[NSIDE]_nu[FREQ].fits`` e.g. ``cib_ns4096_nu0027.fits`` is the map of CIB intensity at 27 GHz and will be used by :py:meth:`.InterpolatingComponent` at that frequency, and can be found on NERSC at `/project/projectdirs/sobs/v4_sims/mbs/websky/data/v0.2`. There are 18 fits files at :math:`N_{side}`=4096 at frequencies [27, 39, 93, 145, 225, 280] +/- 1 GHz. These intensities were selected because in order to be able to interpolate accurately at the 6 frequencies of interest with as few maps as possible. More frequencies will be made available after a full set of map based simulations at those bands that include correlated lensing and SZ components has been implemented.
```
Comment (zonca) [inline docs/models.rst]: @marcelo-alvarez have you also added the HFI frequencies now? also here it still says 0.2

File: docs/gettingstarted.rst (hunk 1 above)
Comment (marcelo-alvarez) [reply docs/gettingstarted.rst]: hope it helps

File: docs/models.rst (hunk 2 above)
Comment (marcelo-alvarez) [reply docs/models.rst]: done

File: docs/models.rst (hunk 3 above)
Comment (marcelo-alvarez) [reply docs/models.rst]: done

File: docs/models.rst (hunk 4 above)
Comment (marcelo-alvarez) [reply docs/models.rst]: done

File: so_pysm_models/tests/test_extragalactic.py (hunk 6 above)
Comment (marcelo-alvarez) [reply so_pysm_models/tests/test_extragalactic.py]: for now this sounds good to me, I will let you know if I think of a better way.

File: docs/models.rst (hunk 7 above)
Comment (marcelo-alvarez) [reply docs/models.rst]: yes I have and also changed to 0.3.

File: so_pysm_models/extragalactic.py (hunk 5 above)
Comment (marcelo-alvarez) [reply so_pysm_models/extragalactic.py]: done.

**Comment by msyriac** [comment]
All of this looks good to me. I think it is ready for merging.

**Comment by zonca** [comment]
Thanks!

**Pull request merged**